### PR TITLE
Prox stepsize

### PR DESCRIPTION
**Pull request by moody-marlin**
Refactored proximal gradient stepsize computation and added test for l1-regularization.

### CODE REVIEW

**Comment by moody-marlin** [comment]
cc @mrocklin 

I think there will be further performance gains with well-placed computes / persists that we haven't done for `proximal_grad` yet, but this pulls out the stepsize computation and adds a test (@hussainsultan ;) )

File: dask_glm/tests/test_regularized.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    return X, y
+
+
+@pytest.mark.parametrize('opt', [proximal_grad])
+@pytest.mark.parametrize('N, p, seed',
+                         [(100, 2, 20009),
+                          (250, 12, 90210),
+                          (95, 6, 70605)])
+def test_l1(N, p, seed, opt):
```
Comment (hussainsultan) [inline dask_glm/tests/test_regularized.py]: nit picky thing - could we change the name of the method from `test_l1` to something more meaningful? Perhaps `test_prox_grad_with_large_alpha`?

**Comment by mrocklin** [comment]
The function fails to run currently (missing reference to `nextBacktrackMult`).

When I remove this reference it does run, but "converges" to values that around 10% apart from each other on repeated runs.

**Comment by moody-marlin** [comment]
Ah this is actually a great catch; for large values of `alpha` we expect `proximal_grad` with `reg = 'l1'` to return a coefficient vector of zeros, but that is *also* the starting value for the algorithm, so I didn't catch the missing reference which appears *after* the convergence check.  Because of this, for this test there should be random initializations.

File: dask_glm/logistic.py [hunk 2]
```diff
@@ -220,6 +220,32 @@ def newton(X, y, max_iter=50, tol=1e-8):
     return beta
 
 
+def compute_prox_step(beta, gradient, func, X, y, prox, stepSize=1.0):
```
Comment (mrocklin) [inline dask_glm/logistic.py]: This function takes in `X` which seems like it would affect scalability.

File: dask_glm/logistic.py [hunk 3]
```diff
@@ -220,6 +220,32 @@ def newton(X, y, max_iter=50, tol=1e-8):
...
+
+    backtrackMult = 0.1
+
+    # Compute the step size
+    obeta = beta
+    lf = func
+    for ii in range(100):
+        beta = prox(obeta - stepSize * gradient, stepSize)
+        step = obeta - beta
+        Xbeta = X.dot(beta).compute()  # ugh
```
Comment (mrocklin) [inline dask_glm/logistic.py]: In bfgs we compute an XStep and then push that forward.  Is that possible here as well?

Comment (moody-marlin) [inline dask_glm/logistic.py]: Unfortunately not; the beta update in proximal gradient descent is non-linear, so the new `Xbeta` can't be computed purely from `obeta` and `Xstep` (at least not that I am aware of).  Maybe @mcg1969 has some input about the best way to do this, avoiding repeated matrix-vector operations within the line search?

Comment (mcg1969) [inline dask_glm/logistic.py]: That's right, but again, the number of backtracks should be small after the first iteration or two.

Comment (mrocklin) [inline dask_glm/logistic.py]: OK, in that case I'll take back my request that we pull this function out.  It seems like all of the line search should happen with Dask and not in a single thread, correct?

**Pull request closed without merging**